### ISSUE TITLE
Android support

### ISSUE THREAD

**Issue by GarboMuffin** [GarboMuffin/p3#2]
Can output Android Studio source code?

**Comment by micahlt** [comment]
Interesting... it seems like this would be hard to do without a virtual machine running `gradle` in the cloud.

**Comment by GarboMuffin** [comment]
The idea would be that it outputs the source code that a user would manually compile, or something like that.

There's also Cordova

**Comment by micahlt** [comment]
Interesting concept!  Cordova, however, still utilizes Gradle under the hood.

**Comment by AwesomeKalin** [comment]
You could output a Cordova project that the user then compiles themselves

**Comment by SurvivCreator** [comment]
@GarboMuffin Have you considered outputing .apk files similar to how website 2 apk outputs .apk from a standalone HTML file.

**Comment by SurvivCreator** [comment]
It's websitetoapk.com I think

**Comment by SurvivCreator** [comment]
Do you think this is possible in the future @GarboMuffin ?

**Comment by GarboMuffin** [comment]
It seems that outputting a valid APK with a signature valid enough to be sideloaded is a very involved process, so that won't be happening any time soon

Telling people to install Android Studio and build it themselves is probably viable

**Comment by hello-smile6** [comment]
> It seems that outputting a valid APK with a signature valid enough to be sideloaded is a very involved process, so that won't be happening any time soon
> 
> Telling people to install Android Studio and build it themselves is probably viable

There's always <a href="https://github.com/copy/v86">v86</a>, could you build in a VM?

**Comment by PPPDUD** [comment]
@GarboMuffin Can we close this now? I feel like this either isn't active enough anymore or it is taking an awful while to consider.

**Comment by GarboMuffin** [comment]
Issues are closed when they've either been fixed, we decide it won't be implemented, etc. We don't close them for just taking a long time.

**Comment by PPPDUD** [comment]
Oh, I see. Good point.

**Comment by PPPDUD** [comment]
Is this in progress still? And, if so, when do you think it will come out?

**Comment by GarboMuffin** [comment]
The volunteer that was working on this (me) happens to have other priorities at the moment

**Comment by originallyxs** [comment]
turbowarp has made on android. email to bcm1668966051@outlook.com to ask for the apk file.

**Comment by originallyxs** [comment]
email us in Chinese or English, to get the file in right language.

**Comment by originallyxs** [comment]
It doesn't seem to be the answer to that question... (garbomuffin seems to ask for turbowarp on android)

**Comment by originallyxs** [comment]
ask us some questions and we'll answer

**Comment by CST1229** [comment]
also why not just upload it anyways without having to email you? (like into a github repo or something)

**Comment by TheShovel** [comment]
> > It doesn't seem to be the answer to that question... (garbomuffin seems to ask for turbowarp on android)
> 
> no, this is very clearly about apk support for *packaged projects*. not turbowarp itself, which works in a browser!

Yeah thats what im talking about... packaging projects as installable apks

**Comment by originallyxs** [comment]
> > > It doesn't seem to be the answer to that question... (garbomuffin seems to ask for turbowarp on android)
> > 
> > no, this is very clearly about apk support for *packaged projects*. not turbowarp itself, which works in a browser!
> 
> Yeah thats what im talking about... packaging projects as installable apks

That's right. I can package html to apk. but you or i must put the html into our repositories.

**Comment by originallyxs** [comment]
where are you?

**Comment by kenzieshane** [comment]
Using Cordova works, I even have exported multiple projects using Cordova already. It's not that hard to figure out—and it works locally without the need to trust anybody else with your project.

**Comment by kenzieshane** [comment]
Can't someone just compile the turbowarp web editor and put the files into cordova or any webview approach?

**Comment by GarboMuffin** [comment]
the hard part of adding android support in the packager is not the part where you load an HTML in a webview. many ways to do that. the hard part is things like generating a fully signed, installable, and updatable APK that will work on all the devices. people will also expect to be able to upload this thing straight to Google Play. all of this has to get done from within a browser where we can't directly run any of Android's standard build tools

**Comment by originallyxs** [comment]
You don't have to upload Google Play,I can convert HTML to APK,But you have to find me.

**Comment by kenzieshane** [comment]
https://blog.kenzieshane.my.id/2025/07/how-to-generate-android-application.html convert HTML to APK by yourself using this, no need to bother anybody

**Comment by kenzieshane** [comment]
Also uhh here's a plug that probably will never get finished because I just do this because I'm bored 

https://github.com/kenzieshane/TurbowarpMobile/

**Comment by GarboMuffin** [comment]
Starting later this year, Google is going to make installing a non-Google Play app substantially more difficult: https://developer.android.com/developer-verification

Either:

- You and *everyone that wants to run your project* has to enable the hidden developer options menu and *wait 24 hours* to enable installing apps from third-party stores, or
- You share your government ID with Google and then actually code sign the app yourself.

Supporting Android in the same way we do Windows/Linux is becoming far more difficult over time, not easier

**Comment by kenzieshane** [comment]
(Oops wrong site, just search for the petition to keep android open)

your voices matter.

**Comment by CST1229** [comment]
> [link redacted]
> 
> your voices matter.

this has more scary banners than a malware site

also no they dont matter

**Comment by ben-jayden** [comment]
Please know it(Keep Android Open) doesn't matter on Chinese Android systems like ColorOS, HarmonyOS, etc. We still can install apks at that time. (Messages from the Web)

**Comment by ben-jayden** [comment]
And you can study CodeMao CoCo(https://coco.codemao.cn) of using Cardova to create an apk

**Comment by GarboMuffin** [comment]
I can't navigate that website because it's in Chinese and it's probably not open source (?) so unlikely there can be any meaningful introspection.

You can use Cordova fine. If putting an HTML file from the packager into Cordova does not work, let me know.

But using Cordova does not really solve any of the problems here. On Android, you still literally have to install Android Studio https://cordova.apache.org/docs/en/latest/guide/platforms/android/index.html and manage keys yourself and everything. It's not clear that this even saves time in the end compared to slapping the HTML into a webview

**Comment by ben-jayden** [comment]
Well, @GarboMuffin There is a next version and it's opened on GitHub(inventocode/coco-next).

**Comment by GarboMuffin** [comment]
Thanks
